### PR TITLE
Feature: Dm mixing

### PR DESCRIPTION
**Pull request by PeizeLin**
The aim of this PR is to achieve issue #2457.

1. add class Mix_Data
2. add class Mix_DMk_2D
3. add DM mixing for Exx_LRI
4. unittests

### CODE REVIEW

File: source/module_io/input.cpp [hunk 1]
```diff
@@ -369,7 +369,7 @@ void Input::Default(void)
 
     exx_separate_loop = true;
     exx_hybrid_step = 100;
-    exx_mixing_beta = 0.0;
+    exx_mixing_beta = 1.0;
```
Comment (maki49) [inline source/module_io/input.cpp:372]: Please modify `module_io/tests`(which failed on CI) corresponding to the change of default value. Thanks~

**Review by maki49 — APPROVED** [review]

**Comment by dyzheng** [comment]
Title of this PR should be modified to "Feature: ...", please add detaild description.

File: source/module_ri/Mix_DMk_2D.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
+//=======================
+// AUTHOR : Peize Lin
+// DATE :   2023-05-09
+//=======================
+
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.cpp:5]: Please add unit tests for new cpp files.

File: source/module_ri/Mix_DMk_2D.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
+//=======================
+// AUTHOR : Peize Lin
+// DATE :   2023-05-09
+//=======================
+
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.h:5]: Please add unit tests for new cpp files.

File: source/module_ri/Mix_Data.hpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
+//=======================
+// AUTHOR : Peize Lin
+// DATE :   2023-05-09
+//=======================
+
+#ifndef MIX_DATA_HPP
+#define MIX_DATA_HPP
```
Comment (dyzheng) [inline source/module_ri/Mix_Data.hpp:7]: Please add unit tests for new hpp files.

File: source/module_esolver/esolver_ks_lcao_elec.cpp [hunk 5]
```diff
@@ -328,6 +328,32 @@ namespace ModuleESolver
 			ModuleBase::timer::tick("ESolver_KS_LCAO", "beforescf");
 			return;
 		}
+		
+		// set initial parameter for mix_DMk_2D
```
Comment (dyzheng) [inline source/module_esolver/esolver_ks_lcao_elec.cpp:332]: please add annotations for this code block, this has changed a file without UT, such as GPT's annotation:
```
This code block checks if a certain flag (cal_exx) is set in a global variable (exx_info.info_global) and configures a 2D mixing object (mix_DMk_2D) accordingly. If the flag is set, it sets the number of k-points in the mixing object to the value in another global variable (kv.nks). If another flag (separate_loop) is set, it checks if a mixing parameter (mixing_beta_for_loop1) is set to 1.0 and sets the mixing mode of the object to "No" if it is, or "Plain" with the mixing parameter value if it is not. If the separate_loop flag is not set, it checks the mixing mode of another global object (CHR_MIX) and sets the mixing mode of the mixing object to "Plain" or "Pulay" accordingly, with the mixing parameter value from CHR_MIX. If the mixing mode is not recognized, it throws an exception with an error message.
```

File: source/module_ri/Mix_DMk_2D.cpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
+//=======================
+// AUTHOR : Peize Lin
+// DATE :   2023-05-09
+//=======================
+
+#include "Mix_DMk_2D.h"
+#include "module_base/tool_title.h"
+
+Mix_DMk_2D &Mix_DMk_2D::set_nks(const int nks)
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.cpp]: Please add annotations for new files, should be doxgen format.

**Review by dyzheng — CHANGES_REQUESTED** [review]

File: source/module_io/input.cpp (hunk 1 above)
Comment (dyzheng) [inline source/module_io/input.cpp:383]: The default value has been changed, how about documentation?

**Comment by maki49** [comment]
@PeizeLin has difficulties in adding unit tests.  (https://github.com/deepmodeling/abacus-develop/issues/2306). I will try to help for this.

File: source/module_ri/Mix_DMk_2D.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+#ifndef MIX_DMK_2D_H
+#define MIX_DMK_2D_H
+
+#include "Mix_Data.h"
+#include "module_base/matrix.h"
+#include "module_base/complexmatrix.h"
+
+#include <vector>
+
+class Mix_DMk_2D
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.h:15]:
```suggestion
/**
 * @class Mix_DMk_2D
 * @brief Class for mixing density matrix in DFT code.
 */
class Mix_DMk_2D
```

File: source/module_ri/Mix_DMk_2D.h [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+{
+public:
+	Mix_DMk_2D &set_nks(const int nks, const bool gamma_only_in);
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.h:24]:
```suggestion
/**
     * @brief Sets the number of k-points and gamma_only flag.
     * @param nks Number of k-points.
     * @param gamma_only_in Flag indicating if only gamma point is mixed.
     * @return Reference to the current object.
     */
	Mix_DMk_2D &set_nks(const int nks, const bool gamma_only_in);
```

File: source/module_ri/Mix_DMk_2D.h [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+	Mix_DMk_2D &set_mixing_mode(const Mixing_Mode mixing_mode);
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.h:30]:
```suggestion
/**
     * @brief Sets the mixing mode.
     * @param mixing_mode Mixing mode.
     * @return Reference to the current object.
     */
	Mix_DMk_2D &set_mixing_mode(const Mixing_Mode mixing_mode);
```

File: source/module_ri/Mix_DMk_2D.h [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    Mix_DMk_2D& set_mixing_beta(const double mixing_beta);
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.h]:
```suggestion
/**
     * @brief Sets the mixing beta.
     * @param mixing_beta Mixing beta.
     * @return Reference to the current object.
     */
    Mix_DMk_2D& set_mixing_beta(const double mixing_beta);
```

File: source/module_ri/Mix_DMk_2D.h [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    template<typename ChgMix>
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.h]:
```suggestion
/**
     * @brief Sets the Pulay mixing coefficient.
     * @tparam ChgMix Type of the charge mixing coefficient.
     * @param iter Iteration number.
     * @param chr_mix Charge mixing coefficient.
     * @return Reference to the current object.
     */
    template<typename ChgMix>
```

File: source/module_ri/Mix_DMk_2D.h [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    Mix_DMk_2D& set_coef_pulay(const int iter, const ChgMix& chr_mix);
+
+	void mix(const std::vector<ModuleBase::matrix> &dm, const bool flag_restart);
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.h:40]:
```suggestion
/**
     * @brief Mixes the density matrix.
     * @param dm double Density matrix.
     * @param flag_restart Flag indicating if it is a restart.
     */
	void mix(const std::vector<ModuleBase::matrix> &dm, const bool flag_restart);

```

File: source/module_ri/Mix_DMk_2D.h [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+	void mix(const std::vector<ModuleBase::ComplexMatrix> &dm, const bool flag_restart);
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.h:47]:
```suggestion
/**
     * @brief Mixes the complex density matrix.
     * @param dm Complex density matrix.
     * @param flag_restart Flag indicating if it is a restart.
     */
	void mix(const std::vector<ModuleBase::ComplexMatrix> &dm, const bool flag_restart);
```

File: source/module_ri/Mix_DMk_2D.h [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+	std::vector<const ModuleBase::matrix*> get_DMk_gamma_out() const;
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.h:53]:
```suggestion
/**
     * @brief Returns the gamma density matrix.
     * @return Vector of pointers to gamma density matrices.
     */
	std::vector<const ModuleBase::matrix*> get_DMk_gamma_out() const;
```

File: source/module_ri/Mix_DMk_2D.h [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+	std::vector<const ModuleBase::ComplexMatrix*> get_DMk_k_out() const;
```
Comment (dyzheng) [inline source/module_ri/Mix_DMk_2D.h:58]:
```suggestion
/**
     * @brief Returns the k-point density matrix.
     * @return Vector of pointers to k-point density matrices.
     */
	std::vector<const ModuleBase::ComplexMatrix*> get_DMk_k_out() const;
```

File: source/module_ri/Mix_Data.hpp [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+//=======================
+
+#ifndef MIX_DATA_HPP
+#define MIX_DATA_HPP
+
+#include "Mix_Data.h"
+#include "module_base/tool_title.h"
+#include "module_elecstate/module_charge/charge_mixing.h"
+
+template<typename Tdata>
```
Comment (dyzheng) [inline source/module_ri/Mix_Data.hpp:18]:
```suggestion
/**
 * @brief Mixes the input data according to the set mixing mode.
 * @param data_in Input data to be mixed.
 * @param flag_restart Flag indicating if it is a restart.
 */
template<typename Tdata>
```

File: source/module_ri/Mix_Data.hpp [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+			break;
+		case Mixing_Mode::Pulay:
+			this->pulay_mixing(data_in, flag_restart);
+			break;
+		default:
+			throw std::domain_error(std::string(__FILE__)+" line "+std::to_string(__LINE__));	break;
+	}	
+}
+
+template<typename Tdata>
```
Comment (dyzheng) [inline source/module_ri/Mix_Data.hpp:42]:
```suggestion
/**
 * @brief Performs Pulay mixing of the input data.
 * @param data_in Input data to be mixed.
 * @param flag_restart Flag indicating if it is a restart.
 */
template<typename Tdata>
```

File: source/module_ri/Mix_Data.hpp [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+		assert(this->data_pulay_list.size() == this->coef_pulay_list.size());
+		assert(this->coef_pulay_list.size() > 0);
+		
+		this->data_out = this->coef_pulay_list[0] * this->data_pulay_list[0];
+		for(std::size_t i=1; i<this->data_pulay_list.size(); ++i)
+			this->data_out += this->coef_pulay_list[i] * this->data_pulay_list[i];
+	}
+}
+
+template<typename Tdata>
```
Comment (dyzheng) [inline source/module_ri/Mix_Data.hpp:74]:
```suggestion
/**
 * @brief Sets the Pulay mixing coefficients.
 * @tparam ChgMix Type of the charge mixing coefficient.
 * @param iter Iteration number.
 * @param chr_mix Charge mixing coefficient.
 */
template<typename Tdata>
```

**Review by dyzheng — APPROVED** [review]
please write detail title and decription for this PR @PeizeLin

**Pull request merged**